File: 之前战队成员的整理/22级战队/ccf cat题解及部分参考代码/The diameter of a rectangle/std.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;

#define N 1000006

int n,root;
int a[N],b[N],L[N],R[N];
ll sum[N];
stack<int >S;
ll ans=0;

void dfs(int x){
    if(!x)return ;
    dfs(L[x]);
    dfs(R[x]);
    sum[x]=sum[L[x]]+sum[R[x]]+1ll*b[x];
}
int main(){
    freopen("20.in","r",stdin);
    freopen("20.out","w",stdout);
    scanf("%d",&n);
    for(int i=1;i<=n;i++)scanf("%d%d",&a[i],&b[i]);
    for(int i=1;i<=n;i++){
        while(!S.empty() && a[S.top()]>a[i]){
            L[i]=S.top();
            S.pop();
        }
        if(!S.empty())R[S.top()]=i;
        S.push(i);
    }
    while(!S.empty()){
        root=S.top();
        S.pop();
    }
    dfs(root);
    for(int i=1;i<=n;i++)ans=max(ans,min(1ll*a[i],sum[i])*min(1ll*a[i],sum[i]));
    cout<<ans<<endl;
}

File: 之前战队成员的整理/22级战队/ccf cat题解及部分参考代码/The diameter of a rectangle/std.py
import sys

sys.setrecursionlimit(10**6)

n, root = 0, 0
N = 1000006
a, b, L, R = [0] * N, [0] * N, [0] * N, [0] * N
sum = [0] * N
S = []
ans = 0

def dfs(x):
    if x == 0:
        return
    dfs(L[x])
    dfs(R[x])
    sum[x] = sum[L[x]] + sum[R[x]] + b[x]

n = int(input())
for i in range(1, n+1):
    a[i], b[i] = map(int, input().split())

for i in range(1, n+1):
    while len(S) > 0 and a[S[-1]] > a[i]:
        L[i] = S[-1]
        S.pop()
    if len(S) > 0:
        R[S[-1]] = i
    S.append(i)

while len(S) > 0:
    root = S[-1]
    S.pop()

dfs(root)

for i in range(1, n+1):
    ans = max(ans, min(a[i], sum[i]) * min(a[i], sum[i]))

print(ans)
